vms/platformvm/state/validator_metadata.go

```diff
@@ -0,0 +1,265 @@
+	DeleteValidatorMetadata(vdrID ids.NodeID, subnetID ids.ID)
+
+	// WriteValidatorMetadata writes all staged updates from prior calls to
+	// SetUptime or SetDelegateeReward.
+	WriteValidatorMetadata(
+		dbPrimary database.KeyValueWriter,
+		dbSubnet database.KeyValueWriter,
+	) error
+}
+
+type metadata struct {
+	metadata map[ids.NodeID]map[ids.ID]*validatorMetadata // vdrID -> subnetID -> metadata
+	// updatedMetadata tracks the updates since WriteValidatorMetadata was last called
+	updatedMetadata map[ids.NodeID]set.Set[ids.ID] // vdrID -> subnetIDs
+}
+
+func newValidatorState() validatorState {
+	return &metadata{
+		metadata:        make(map[ids.NodeID]map[ids.ID]*validatorMetadata),
+		updatedMetadata: make(map[ids.NodeID]set.Set[ids.ID]),
+	}
+}
+
+func (m *metadata) LoadValidatorMetadata(
+	vdrID ids.NodeID,
+	subnetID ids.ID,
+	uptime *validatorMetadata,
+) {
+	subnetMetadata, ok := m.metadata[vdrID]
+	if !ok {
+		subnetMetadata = make(map[ids.ID]*validatorMetadata)
+		m.metadata[vdrID] = subnetMetadata
+	}
+	subnetMetadata[subnetID] = uptime
+}
+
+func (m *metadata) GetUptime(
+	vdrID ids.NodeID,
+	subnetID ids.ID,
+) (time.Duration, time.Time, error) {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return 0, time.Time{}, database.ErrNotFound
+	}
+	return metadata.UpDuration, metadata.lastUpdated, nil
+}
+
+func (m *metadata) SetUptime(
+	vdrID ids.NodeID,
+	subnetID ids.ID,
+	upDuration time.Duration,
+	lastUpdated time.Time,
+) error {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return database.ErrNotFound
+	}
+	metadata.UpDuration = upDuration
+	metadata.lastUpdated = lastUpdated
+
+	m.addUpdatedMetadata(vdrID, subnetID)
+	return nil
+}
+
+func (m *metadata) GetDelegateeReward(
+	subnetID ids.ID,
+	vdrID ids.NodeID,
+) (uint64, error) {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return 0, database.ErrNotFound
+	}
+	return metadata.PotentialDelegateeReward, nil
+}
+
+func (m *metadata) SetDelegateeReward(
+	subnetID ids.ID,
+	vdrID ids.NodeID,
+	amount uint64,
+) error {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return database.ErrNotFound
+	}
+	metadata.PotentialDelegateeReward = amount
+
+	m.addUpdatedMetadata(vdrID, subnetID)
+	return nil
+}
+
+func (m *metadata) DeleteValidatorMetadata(vdrID ids.NodeID, subnetID ids.ID) {
+	subnetMetadata := m.metadata[vdrID]
+	delete(subnetMetadata, subnetID)
+	if len(subnetMetadata) == 0 {
+		delete(m.metadata, vdrID)
+	}
+
+	subnetUpdatedMetadata := m.updatedMetadata[vdrID]
+	subnetUpdatedMetadata.Remove(subnetID)
+	if subnetUpdatedMetadata.Len() == 0 {
+		delete(m.updatedMetadata, vdrID)
+	}
+}
+
+func (m *metadata) WriteValidatorMetadata(
+	dbPrimary database.KeyValueWriter,
+	dbSubnet database.KeyValueWriter,
+) error {
+	for vdrID, updatedSubnets := range m.updatedMetadata {
+		for subnetID := range updatedSubnets {
+			metadata := m.metadata[vdrID][subnetID]
```

Comment:
I know we document in `state.go` that this has to be called after `writeCurrentStakers` but should we note that here as well?

Reply:
That seems like a pretty clear abstraction break no? Like we could say `WriteValidatorMetadata` should be called after all of the calls to `SetUptime`, `SetDelegateeReward`, and `DeleteValidatorMetadata`... But that seems pretty obvious imo.